### PR TITLE
mgr/cephadm: save host cache data after scheduling daemon action

### PR DESCRIPTION
**Pull request by adk3798**
Otherwise, the actions could be lost on a mgr failover despite us including shceduled actions as part of the data we store in the config-key entry for the host.

This also makes it impossible to redeploy the active mgr with a new image in one command.

Fixes: https://tracker.ceph.com/issues/57884

Signed-off-by: Adam King <adking@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/module.py [hunk 1]
```diff
@@ -2148,6 +2148,7 @@ def _schedule_daemon_action(self, daemon_name: str, action: str) -> str:
             raise OrchestratorError(
                 f'Unable to schedule redeploy for {daemon_name}: No standby MGRs')
         self.cache.schedule_daemon_action(dd.hostname, dd.name(), action)
+        self.cache.save_host(dd.hostname)
```
Comment (mgfritch) [inline src/pybind/mgr/cephadm/module.py:2151]: assert `save_host` was called during the unit test?

```patch
diff --git a/src/pybind/mgr/cephadm/tests/test_cephadm.py b/src/pybind/mgr/cephadm/tests/test_cephadm.py
index 76a1073fc4a..1b2ba7ec494 100644
--- a/src/pybind/mgr/cephadm/tests/test_cephadm.py
+++ b/src/pybind/mgr/cephadm/tests/test_cephadm.py
@@ -404,13 +404,15 @@ class TestCephadm(object):
         ]
     )
     @mock.patch("cephadm.serve.CephadmServe._run_cephadm", _run_cephadm('{}'))
-    def test_daemon_check(self, cephadm_module: CephadmOrchestrator, action):
+    @mock.patch("cephadm.module.HostCache.save_host")
+    def test_daemon_check(self, save_host, cephadm_module: CephadmOrchestrator, action):
         with with_host(cephadm_module, 'test'):
             with with_service(cephadm_module, ServiceSpec(service_type='grafana'), CephadmOrchestrator.apply_grafana, 'test') as d_names:
                 [daemon_name] = d_names
 
                 cephadm_module._schedule_daemon_action(daemon_name, action)
 
+                save_host.assert_called_with('test')
                 assert cephadm_module.cache.get_scheduled_daemon_action(
                     'test', daemon_name) == action
```

**Review by mgfritch — APPROVED** [review]
update the unit test? otherwise lgtm!

File: src/pybind/mgr/cephadm/module.py (hunk 1 above)
Comment (adk3798) [reply src/pybind/mgr/cephadm/module.py:2151]: done

**Comment by adk3798** [comment]
https://pulpito.ceph.com/adking-2022-12-02_21:56:44-orch:cephadm-wip-adk-testing-2022-12-02-0932-distro-default-smithi/

Failures tracked by:

- https://tracker.ceph.com/issues/49287
- https://tracker.ceph.com/issues/57755
- https://tracker.ceph.com/issues/57771
- https://tracker.ceph.com/issues/58145
- https://tracker.ceph.com/issues/58173

It's a lot of failures, but other than the the last one, they are all things we were aware of before the run (even the dead mgr-nfs-upgrade job had a `mount.nfs: mounting smithi174:/fake failed, reason given by server: No such file or directory` issue). The last one (https://tracker.ceph.com/issues/58173) as well seems to be a failure on a test setting flags on a pool, and given all the PRs in this run strictly touched mgr modules, I think it's unlikely any of the attached PRs caused the failures. Overall, I think the failures should only block merging on PRs related to nfs due to https://tracker.ceph.com/issues/58145 making it so we can't properly run our nfs related tests currently.

**Comment by adk3798** [comment]
jenkins retest this please

**Comment by adk3798** [comment]
jenkins test make check

**Pull request merged**